arctic/store/metadata_store.py

```diff
@@ -57,8 +59,50 @@ def __repr__(self):
         return str(self)
 
     @mongo_retry
-    def list_symbols(self):
-        return self.distinct('symbol')
+    def list_symbols(self, regex=None, as_of=None, **kwargs):
+        """
+         Return the symbols in this library.
+
+         Parameters
+         ----------
+         as_of : `datetime.datetime`
+            filter symbols valid at given time
+         regex : `str`
+             filter symbols by the passed in regular expression
+         kwargs :
+             kwarg keys are used as fields to query for symbols with metadata matching
+             the kwargs query
+
+         Returns
+         -------
+         String list of symbols in the library
+        """
+        query = {}
+        if regex is not None:
+            query['symbol'] = {'$regex': regex}
+
+        if as_of is not None:
+            query['start_time'] = {'$lte': as_of}
+
+        if kwargs:
+            for k, v in six.iteritems(kwargs):
+                query['metadata.' + k] = v
+
+        if len(query):
+            pipeline = []
+            if query:
+                # Match based on user criteria first
+                pipeline.append({'$match': query})
+
+            pipeline.extend([
+                {'$sort': {'start_time': pymongo.DESCENDING}},
+                # Group by 'symbol'
+                {'$group': {'_id': '$symbol'}},
```

Comment:
Can we use $max on the start_time here instead of sorting? https://docs.mongodb.com/manual/reference/operator/aggregation/max/#grp._S_max

The current sort is not hitting the index (('symbol', 1), ('start_time', -1)) so it will do a full scan, sorting all documents in the collection every time this is called. If we group by start_time we are still going over the entire collection, but at least at a cost of O(N)